File: frontend/src/schema/equipment.ts
/**
 * Equipment taxonomy for P&ID canvases.
 *
 * `EquipmentType` is the stable key used everywhere. `EQUIPMENT_METADATA` carries
 * rendering + port info shared by the Vue symbol layer and the AI agents, so
 * there is one place that defines "what a pump is". Symbols are ISA/ISO-styled
 * SVGs under `src/assets/symbols/`, keyed by `SymbolKey`.
 */

export const EQUIPMENT_TYPES = [
  // pumps
  'centrifugal_pump',
  'positive_displacement_pump',
  // compressors / blowers
  'compressor',
  'blower',
  'fan',
  // drivers
  'motor',
  'turbine',
  // valves
  'gate_valve',
  'globe_valve',
  'ball_valve',
  'butterfly_valve',
  'check_valve',
  'control_valve',
  'three_way_valve',
  'relief_valve',
  // vessels
  'vessel',
  'column',
  'reactor',
  'mixer',
  // tanks
  'storage_tank',
  // exchangers
  'shell_tube_heat_exchanger',
  'air_cooler',
  'cooling_tower',
  // separation
  'filter',
  // instrument
  'instrument',
] as const;

export type EquipmentType = (typeof EQUIPMENT_TYPES)[number];

/** SVG symbol keys (one per file in assets/symbols). */
export type SymbolKey =
  | 'pump_centrifugal'
  | 'pump_pd'
  | 'compressor'
  | 'blower'
  | 'motor'
  | 'turbine'
  | 'valve_gate'
  | 'valve_globe'
  | 'valve_ball'
  | 'valve_butterfly'
  | 'valve_check'
  | 'valve_control'
  | 'valve_threeway'
  | 'valve_relief'
  | 'vessel'
  | 'column'
  | 'reactor'
  | 'mixer'
  | 'tank'
  | 'heat_exchanger'
  | 'air_cooler'
  | 'cooling_tower'
  | 'filter'
  | 'instrument';

export type PortRole = 'inlet' | 'outlet' | 'inout';
export type PortSide = 'top' | 'right' | 'bottom' | 'left';

export interface PortDef {
  id: string;
  role: PortRole;
  side: PortSide;
  label?: string;
}

export type EquipmentCategory =
  | 'pump'
  | 'compressor'
  | 'driver'
  | 'valve'
  | 'vessel'
  | 'tank'
  | 'exchanger'
  | 'separation'
  | 'instrument';

export interface EquipmentMeta {
  type: EquipmentType;
  label: string;
  category: EquipmentCategory;
  symbol: SymbolKey;
  size: { width: number; height: number };
  ports: PortDef[];
}

const IN_OUT: PortDef[] = [
  { id: 'in', role: 'inlet', side: 'left' },
  { id: 'out', role: 'outlet', side: 'right' },
];

function meta(
  type: EquipmentType,
  label: string,
  category: EquipmentCategory,
  symbol: SymbolKey,
  width: number,
  height: number,
  ports: PortDef[],
): EquipmentMeta {
  return { type, label, category, symbol, size: { width, height }, ports };
}

export const EQUIPMENT_METADATA: Record<EquipmentType, EquipmentMeta> = {
  centrifugal_pump: meta(
    'centrifugal_pump',
    'Centrifugal Pump',
    'pump',
    'pump_centrifugal',
    80,
    80,
    [
      { id: 'suction', role: 'inlet', side: 'left' },
      { id: 'discharge', role: 'outlet', side: 'top' },
    ],
  ),
  positive_displacement_pump: meta(
    'positive_displacement_pump',
    'Positive Displacement Pump',
    'pump',
    'pump_pd',
    80,
    80,
    [
      { id: 'suction', role: 'inlet', side: 'left' },
      { id: 'discharge', role: 'outlet', side: 'right' },
    ],
  ),
  compressor: meta('compressor', 'Compressor', 'compressor', 'compressor', 92, 80, [
    { id: 'suction', role: 'inlet', side: 'left' },
    { id: 'discharge', role: 'outlet', side: 'right' },
  ]),
  blower: meta('blower', 'Blower', 'compressor', 'blower', 82, 80, IN_OUT),
  fan: meta('fan', 'Fan', 'compressor', 'blower', 80, 80, IN_OUT),
  motor: meta('motor', 'Motor', 'driver', 'motor', 70, 70, [
    { id: 'shaft', role: 'inout', side: 'right' },
  ]),
  turbine: meta('turbine', 'Turbine / Expander', 'driver', 'turbine', 92, 72, IN_OUT),
  gate_valve: meta('gate_valve', 'Gate Valve', 'valve', 'valve_gate', 64, 40, IN_OUT),
  globe_valve: meta('globe_valve', 'Globe Valve', 'valve', 'valve_globe', 64, 40, IN_OUT),
  ball_valve: meta('ball_valve', 'Ball Valve', 'valve', 'valve_ball', 64, 40, IN_OUT),
  butterfly_valve: meta(
    'butterfly_valve',
    'Butterfly Valve',
    'valve',
    'valve_butterfly',
    64,
    40,
    IN_OUT,
  ),
  check_valve: meta('check_valve', 'Check Valve', 'valve', 'valve_check', 64, 40, IN_OUT),
  control_valve: meta('control_valve', 'Control Valve', 'valve', 'valve_control', 64, 58, IN_OUT),
  three_way_valve: meta('three_way_valve', 'Three-Way Valve', 'valve', 'valve_threeway', 56, 56, [
    { id: 'in', role: 'inlet', side: 'left' },
    { id: 'out', role: 'outlet', side: 'right' },
    { id: 'branch', role: 'outlet', side: 'bottom' },
  ]),
  relief_valve: meta('relief_valve', 'Relief Valve', 'valve', 'valve_relief', 54, 66, [
    { id: 'in', role: 'inlet', side: 'bottom' },
    { id: 'out', role: 'outlet', side: 'left' },
  ]),
  vessel: meta('vessel', 'Vessel / Drum', 'vessel', 'vessel', 100, 140, [
    { id: 'top', role: 'inout', side: 'top' },
    { id: 'bottom', role: 'outlet', side: 'bottom' },
    { id: 'side_in', role: 'inlet', side: 'left' },
    { id: 'side_out', role: 'outlet', side: 'right' },
  ]),
  column: meta('column', 'Column / Tower', 'vessel', 'column', 92, 200, [
    { id: 'overhead', role: 'outlet', side: 'top' },
    { id: 'bottoms', role: 'outlet', side: 'bottom' },
    { id: 'feed', role: 'inlet', side: 'left' },
  ]),
  reactor: meta('reactor', 'Reactor', 'vessel', 'reactor', 110, 150, [
    { id: 'feed', role: 'inlet', side: 'top' },
    { id: 'product', role: 'outlet', side: 'bottom' },
    { id: 'jacket_in', role: 'inlet', side: 'left' },
    { id: 'jacket_out', role: 'outlet', side: 'right' },
  ]),
  mixer: meta('mixer', 'Mixer / Agitated Tank', 'vessel', 'mixer', 112, 120, [
    { id: 'in', role: 'inlet', side: 'left' },
    { id: 'out', role: 'outlet', side: 'bottom' },
  ]),
  storage_tank: meta('storage_tank', 'Storage Tank', 'tank', 'tank', 140, 116, [
    { id: 'fill', role: 'inlet', side: 'top' },
    { id: 'draw', role: 'outlet', side: 'bottom' },
  ]),
  shell_tube_heat_exchanger: meta(
    'shell_tube_heat_exchanger',
    'Shell & Tube Heat Exchanger',
    'exchanger',
    'heat_exchanger',
    120,
    80,
    [
      { id: 'shell_in', role: 'inlet', side: 'left' },
      { id: 'shell_out', role: 'outlet', side: 'right' },
      { id: 'tube_in', role: 'inlet', side: 'top' },
      { id: 'tube_out', role: 'outlet', side: 'bottom' },
    ],
  ),
  air_cooler: meta('air_cooler', 'Air Cooler', 'exchanger', 'air_cooler', 120, 84, IN_OUT),
  cooling_tower: meta('cooling_tower', 'Cooling Tower', 'exchanger', 'cooling_tower', 120, 110, [
    { id: 'in', role: 'inlet', side: 'top' },
    { id: 'out', role: 'outlet', side: 'bottom' },
  ]),
  filter: meta('filter', 'Filter', 'separation', 'filter', 72, 96, [
    { id: 'in', role: 'inlet', side: 'top' },
    { id: 'out', role: 'outlet', side: 'bottom' },
  ]),
  instrument: meta('instrument', 'Instrument', 'instrument', 'instrument', 56, 56, [
    { id: 'signal', role: 'inout', side: 'bottom' },
  ]),
};

export function isEquipmentType(value: string): value is EquipmentType {
  return (EQUIPMENT_TYPES as readonly string[]).includes(value);
}

export function getEquipmentMeta(type: EquipmentType): EquipmentMeta {
  return EQUIPMENT_METADATA[type];
}

File: backend/src/inf_canvas/schema/equipment.py
"""Equipment taxonomy + port metadata (Python mirror of equipment.ts).

The agents rely on `EQUIPMENT_METADATA` to know valid equipment types, their
canvas footprint, and which ports exist (used for connection routing). Keep in
lockstep with frontend/src/schema/equipment.ts.
"""

from typing import Literal

from pydantic import BaseModel

EquipmentType = Literal[
    "centrifugal_pump",
    "positive_displacement_pump",
    "compressor",
    "blower",
    "fan",
    "motor",
    "turbine",
    "gate_valve",
    "globe_valve",
    "ball_valve",
    "butterfly_valve",
    "check_valve",
    "control_valve",
    "three_way_valve",
    "relief_valve",
    "vessel",
    "column",
    "reactor",
    "mixer",
    "storage_tank",
    "shell_tube_heat_exchanger",
    "air_cooler",
    "cooling_tower",
    "filter",
    "instrument",
]

EQUIPMENT_TYPES: tuple[EquipmentType, ...] = (
    "centrifugal_pump",
    "positive_displacement_pump",
    "compressor",
    "blower",
    "fan",
    "motor",
    "turbine",
    "gate_valve",
    "globe_valve",
    "ball_valve",
    "butterfly_valve",
    "check_valve",
    "control_valve",
    "three_way_valve",
    "relief_valve",
    "vessel",
    "column",
    "reactor",
    "mixer",
    "storage_tank",
    "shell_tube_heat_exchanger",
    "air_cooler",
    "cooling_tower",
    "filter",
    "instrument",
)

PortRole = Literal["inlet", "outlet", "inout"]
PortSide = Literal["top", "right", "bottom", "left"]


class PortDef(BaseModel):
    id: str
    role: PortRole
    side: PortSide
    label: str | None = None


class Size(BaseModel):
    width: float
    height: float


class EquipmentMeta(BaseModel):
    type: EquipmentType
    label: str
    symbol: str
    size: Size
    ports: list[PortDef]


def _meta(
    type_: EquipmentType,
    label: str,
    symbol: str,
    width: float,
    height: float,
    ports: list[tuple[str, PortRole, PortSide]],
) -> EquipmentMeta:
    return EquipmentMeta(
        type=type_,
        label=label,
        symbol=symbol,
        size=Size(width=width, height=height),
        ports=[PortDef(id=p[0], role=p[1], side=p[2]) for p in ports],
    )


_IN_OUT: list[tuple[str, PortRole, PortSide]] = [
    ("in", "inlet", "left"),
    ("out", "outlet", "right"),
]

EQUIPMENT_METADATA: dict[EquipmentType, EquipmentMeta] = {
    "centrifugal_pump": _meta(
        "centrifugal_pump",
        "Centrifugal Pump",
        "pump_centrifugal",
        80,
        80,
        [("suction", "inlet", "left"), ("discharge", "outlet", "top")],
    ),
    "positive_displacement_pump": _meta(
        "positive_displacement_pump",
        "Positive Displacement Pump",
        "pump_pd",
        80,
        80,
        [("suction", "inlet", "left"), ("discharge", "outlet", "right")],
    ),
    "compressor": _meta(
        "compressor",
        "Compressor",
        "compressor",
        92,
        80,
        [("suction", "inlet", "left"), ("discharge", "outlet", "right")],
    ),
    "blower": _meta("blower", "Blower", "blower", 82, 80, _IN_OUT),
    "fan": _meta("fan", "Fan", "blower", 80, 80, _IN_OUT),
    "motor": _meta("motor", "Motor", "motor", 70, 70, [("shaft", "inout", "right")]),
    "turbine": _meta("turbine", "Turbine / Expander", "turbine", 92, 72, _IN_OUT),
    "gate_valve": _meta("gate_valve", "Gate Valve", "valve_gate", 64, 40, _IN_OUT),
    "globe_valve": _meta("globe_valve", "Globe Valve", "valve_globe", 64, 40, _IN_OUT),
    "ball_valve": _meta("ball_valve", "Ball Valve", "valve_ball", 64, 40, _IN_OUT),
    "butterfly_valve": _meta(
        "butterfly_valve", "Butterfly Valve", "valve_butterfly", 64, 40, _IN_OUT
    ),
    "check_valve": _meta("check_valve", "Check Valve", "valve_check", 64, 40, _IN_OUT),
    "control_valve": _meta("control_valve", "Control Valve", "valve_control", 64, 58, _IN_OUT),
    "three_way_valve": _meta(
        "three_way_valve",
        "Three-Way Valve",
        "valve_threeway",
        56,
        56,
        [("in", "inlet", "left"), ("out", "outlet", "right"), ("branch", "outlet", "bottom")],
    ),
    "relief_valve": _meta(
        "relief_valve",
        "Relief Valve",
        "valve_relief",
        54,
        66,
        [("in", "inlet", "bottom"), ("out", "outlet", "left")],
    ),
    "vessel": _meta(
        "vessel",
        "Vessel / Drum",
        "vessel",
        100,
        140,
        [
            ("top", "inout", "top"),
            ("bottom", "outlet", "bottom"),
            ("side_in", "inlet", "left"),
            ("side_out", "outlet", "right"),
        ],
    ),
    "column": _meta(
        "column",
        "Column / Tower",
        "column",
        92,
        200,
        [
            ("overhead", "outlet", "top"),
            ("bottoms", "outlet", "bottom"),
            ("feed", "inlet", "left"),
        ],
    ),
    "reactor": _meta(
        "reactor",
        "Reactor",
        "reactor",
        110,
        150,
        [
            ("feed", "inlet", "top"),
            ("product", "outlet", "bottom"),
            ("jacket_in", "inlet", "left"),
            ("jacket_out", "outlet", "right"),
        ],
    ),
    "mixer": _meta(
        "mixer",
        "Mixer / Agitated Tank",
        "mixer",
        112,
        120,
        [("in", "inlet", "left"), ("out", "outlet", "bottom")],
    ),
    "storage_tank": _meta(
        "storage_tank",
        "Storage Tank",
        "tank",
        140,
        116,
        [("fill", "inlet", "top"), ("draw", "outlet", "bottom")],
    ),
    "shell_tube_heat_exchanger": _meta(
        "shell_tube_heat_exchanger",
        "Shell & Tube Heat Exchanger",
        "heat_exchanger",
        120,
        80,
        [
            ("shell_in", "inlet", "left"),
            ("shell_out", "outlet", "right"),
            ("tube_in", "inlet", "top"),
            ("tube_out", "outlet", "bottom"),
        ],
    ),
    "air_cooler": _meta("air_cooler", "Air Cooler", "air_cooler", 120, 84, _IN_OUT),
    "cooling_tower": _meta(
        "cooling_tower",
        "Cooling Tower",
        "cooling_tower",
        120,
        110,
        [("in", "inlet", "top"), ("out", "outlet", "bottom")],
    ),
    "filter": _meta(
        "filter",
        "Filter",
        "filter",
        72,
        96,
        [("in", "inlet", "top"), ("out", "outlet", "bottom")],
    ),
    "instrument": _meta(
        "instrument", "Instrument", "instrument", 56, 56, [("signal", "inout", "bottom")]
    ),
}
